datasets/de/abgeordnetenwatch/crawler.py

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import json
+import math
+from zavod import Context
+from zavod import helpers as h
+
+
+def crawl(context: Context):
+    # Fetch the source data URL specified in the metadata to a local path:
+    api_response = context.fetch_json(context.dataset.data.url, cache_days=30)
+    total_results = api_response["meta"]["result"]["total"]
+    num_batches = math.ceil(total_results / 100)
+
+    bi = 0
+    while bi < num_batches:
+        context.log.info(
+            "Get Politicians from range {} to {}".format(bi * 100, (bi + 1) * 100)
+        )
+        if bi > 0:
+            api_response = context.fetch_json(
+                context.dataset.data.url,
+                params={"range_start": bi * 100},
+                cache_days=30,
+            )
+            batch_total_results = api_response["meta"]["result"]["total"]
+            if batch_total_results != total_results:
+                num_batches = math.ceil(batch_total_results / 100)
+
+        for candidate in api_response.get("data"):
+            parliament_info = candidate["parliament_period"]["label"].split(" ")
+            state = parliament_info[0]
+            inception_date = parliament_info[1]
+            dissolution_date = parliament_info[3]
+
+            position = h.make_position(
+                context,
+                name="Member of Parliament",
+                country="Germany",
+                subnational_area=state,
+                inception_date=inception_date,
+                dissolution_date=dissolution_date,
+            )
+
+            person = context.make("Person")
+            person.id = context.make_id(candidate["politician"]["id"])
+
+            # Get the politician birthday date
+            context.log.info(
+                "Get Politician {} birth year".format(candidate["politician"]["label"])
+            )
+            api_response = context.fetch_json(
+                candidate["politician"]["api_url"], cache_days=30
+            )
+            birthYear = api_response["data"]["year_of_birth"]
+            person.add("birthDate", birthYear)
+
+            # Get fullname
+            h.apply_name(person, full=candidate["politician"]["label"])
+            occupancy = h.make_occupancy(
+                context, person, position, False, start_date=candidate["start_date"]
```

Comment:
Consider using candidate.pop to access its members, and then context.audit(candidate) so that we automatically get warnings if new or unexpected fields show up, e.g. if they add end_dates. https://zavod.opensanctions.org/patterns/#detect-unhandled-data

Reply:
Just did it both suggestions. I'm not sure if I understood it correctly how the context.audit_data works because now I'm having a log of warnings from "Unexpected data found".

Reply:
Aha - I wasn't very clear.

Use `context.audit` once you've popped everything you feel you need from the dict. At that point you can see what fields remain and decide one by one whether you want its data.

e.g. after the following change I see there's an abgeordnetenwatch_url which would go nicely in Person.sourceUrl:

```diff
--- a/datasets/de/abgeordnetenwatch/crawler.py
+++ b/datasets/de/abgeordnetenwatch/crawler.py
@@ -9,7 +9,6 @@ from zavod import helpers as h
 def get_politician_data(context: Context, politician_api_url: str):
     api_response = context.fetch_json(politician_api_url, cache_days=30)
     politician = api_response.pop("data")
-    context.audit_data(politician)
     return politician
 
 
@@ -38,7 +37,6 @@ def crawl(context: Context):
                 num_batches = math.ceil(batch_total_results / 100)
 
         for mandate in api_response.get("data"):
-            context.audit_data(mandate)
             politician = mandate.pop("politician")
 
             # Get state from parliament info
@@ -57,6 +55,8 @@ def crawl(context: Context):
             person.id = context.make_id(politician.pop("id"))
             politician_fullname = politician.pop("label")
 
+            # abgeordnetenwatch_url -> sourceUrl
+
             # Get the politician birthday date
             context.log.info("Get Politician {} birth year".format(politician_fullname))
             politician_data = get_politician_data(context, politician.pop("api_url"))
@@ -75,5 +75,17 @@ def crawl(context: Context):
                 context.emit(position)
                 context.emit(occupancy)
 
+            context.audit_data(mandate, ignore=[
+                "id",
+                "entity_type",
+                "label",
+                "api_url",
+                "type",
+                "electoral_data",
+                "electoral_list",
+                "fraction_membership",
+            ])
+            context.audit_data(politician, ignore=["entity_type"])
+
```

Reply:
Thanks for clarifying, just put all the fields that I'm not using with pop on ignore and now running the crawler I don't have any unexpected field warning anymore.